commands/generate.go

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Alex Ellis 2017. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package commands
+
+import (
+	"fmt"
+
+	"github.com/openfaas/faas-cli/schema"
+	"github.com/openfaas/faas-cli/stack"
+	"github.com/spf13/cobra"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	functionNamespace = "openfaas-fn"
+	resourceKind      = "Function"
+)
+
+var api string
+
+func init() {
+	generateCmd.Flags().StringVar(&api, "api", "", "OpenFaaS CRD API version")
+	faasCmd.AddCommand(generateCmd)
+}
+
+var generateCmd = &cobra.Command{
+	Use:   "generate --api=openfaas.com/v1alpha2 --yaml stack.yml",
+	Short: "Generate Kubernetes CRD YAML file",
+	Long:  `The generate command creates kubernetes CRD YAML file for functions`,
+	Example: `faas-cli generate --api=openfaas.com/v1alpha2 --yaml stack.yml | kubectl apply  -f -
+faas-cli generate --api=openfaas.com/v1alpha2 -f stack.yml`,
+	PreRunE: preRunGenerate,
+	RunE:    runGenerate,
+}
+
+func preRunGenerate(cmd *cobra.Command, args []string) error {
+	if len(api) == 0 {
+		return fmt.Errorf("You must supply api version with the --api flag")
+	}
+	return nil
+}
+
+func runGenerate(cmd *cobra.Command, args []string) error {
+
+	var services stack.Services
+
+	//Process function stack file
+	if len(yamlFile) > 0 {
+		parsedServices, err := stack.ParseYAMLFile(yamlFile, regex, filter)
+		if err != nil {
+			return err
+		}
+
+		if parsedServices != nil {
+			services = *parsedServices
+		}
+	}
+	var objectsString string
+
+	if len(services.Functions) > 0 {
+		for name, function := range services.Functions {
+			//read environment variables from the file
+			fileEnvironment, err := readFiles(function.EnvironmentFile)
+			if err != nil {
+				return err
+			}
+
+			//combine all environment variables
+			allEnvironment, envErr := compileEnvironment([]string{}, function.Environment, fileEnvironment)
+			if envErr != nil {
+				return envErr
+			}
+
+			metadata := schema.Metadata{Name: name, Namespace: functionNamespace}
+			spec := schema.Spec{
+				Name:        name,
+				Image:       function.Image,
+				Environment: allEnvironment,
+				Labels:      function.Labels,
+				Limits:      function.Limits,
+				Requests:    function.Requests,
+				Constraints: function.Constraints,
+				Secrets:     function.Secrets,
+			}
+
+			crd := schema.CRD{
+				APIVersion: api,
+				Kind:       resourceKind,
+				Metadata:   metadata,
+				Spec:       spec,
+			}
+
+			//Marshal the object definition to yaml
+			objectString, err := yaml.Marshal(crd)
+			if err != nil {
+				return err
+			}
+			objectsString += string(objectString) + "\n---\n"
```

Comment:
The convention is to put the `---` at the beginning of the definition not at the end:

```
objectsString += "---\n" +  string(objectString) 
```